Upgrader - Add support for automatic snapshots

totten opened the pull request:
Overview
----------------------------------------

CiviCRM upgrades sometimes manipulate MySQL data. However, any time you change the data, there is a risk of changing it incorrectly. This PR adds a general snapshot/backup mechanism for use in upgrade scripts, eg

```php
public function upgrade_N_N_N($rev) {
  $this->addSnapshotTask('my_snapshot', CRM_Utils_SQL_Select::from('civicrm_foobar')
    ->select('id, whiz, bang')
  );
}
```

which creates a snapshot-table similar to this:

```
mysql> SELECT * FROM snap_civicrm_v5_50_my_snapshot;
+----+------+-------------+
| id | whiz | bang        |
+----+------+-------------+
| 1  | aaa  | eh eh eh    |
| 2  | bbb  | bee bee bee |
| 3  | ccc  | see see see |
+----+------+-------------+
```

(*Note: During discussion of the PR, the naming formula switched; it was `civicrm_snap_v5_50_my_snapshot`
and became `snap_civicrm_v5_50_my_snapshot`.*)

Motivation
----------------------------------------

In recent months, we had two releases where the upgrade script caused lossy changes that were initially difficult to discover - and subsequently difficult to cleanup (eg [event timezones/timestamps](https://lab.civicrm.org/dev/core/-/wikis/CiviEvent-v5.47-Timezone-Notice) and [reminders/`limit_to`](https://lab.civicrm.org/dev/core/-/wikis/CiviCRM-v5.49-Schedule-Reminder-Notice)).

In both cases, we were able to develop mitigations that *probably-mostly* restored the data, but they required a bunch of effort (*eg working backwards from flawed event-times and guessing about the user's DST*), and they still couldn't fix everything, so they required publishing advisories to describe the problematic-cases (*where manual cleanup is necessary*).

Adding snapshot support directly to the upgrade is a fairly simple change. It would significantly reduce the guess-work involved with restoring munged data, and (for a large number of Civi sites) the cost of storing some extra snapshots should be quite acceptable.

Technical Details
----------------------------------------

As a first+simple example, the PR incorporates a snapshot at the start of the `5.50.alpha1` upgrade:

```php
public function upgrade_5_50_alpha1($rev): void {
    $this->addSnapshotTask('mappings', CRM_Utils_SQL_Select::from('civicrm_mapping'));
    $this->addSnapshotTask('fields', CRM_Utils_SQL_Select::from('civicrm_mapping_field'));
```

This runs two queries and stores two snapshot-tables:

*  `SELECT * FROM civicrm_mapping` ==> `snap_civicrm_v5_50_mappings`
*  `SELECT * FROM civicrm_mapping_field` ==> `snap_civicrm_v5_50_fields`

Here's a more interesting example -- a hypothetical revision to `display_name`:

```php
function upgrade_5_60_alpha1$($rev): void {
  $this->addSnapshotTask('names', CRM_Utils_SQL_Select::from('civicrm_contact')
    ->select('id,display_name'));
  $this->addTask(ts('Drop old version of "display_name"', 'dropColumn', 'civicrm_contact', 'display_name'));
  $this->addTask(ts('Re-create display_name as generated column', 'addColumn', 'display_name'
    'VARCHAR GENERATED ALWAYS AS (CONCAT(first_name, ' ', last_name)) STORED');
}
```

This also creates a snapshot-table:

* `SELECT id, display_name FROM civicrm_contact` ==> `snap_civicrm_v5_60_names`

This example is interesting for a couple reasons:

* The `civicrm_contact` table has many columns that don't change. We only need to backup the one that changes.
* The `civicrm_contact` table can be large --  large enough to merit pagination. The `addSnapshotTask()` will paginate the work. (It currently runs with pages of 50k rows.)

Both examples create MySQL tables with predictable names (`snap_civicrm_{$VERSION}_{$NAME}`). The tables can be cleaned up in a few ways:

* Future upgrades will automatically remove old snapshots after +4 versions. (Thus, when upgrading to 5.64, it will drop `snap_civicrm_v5_60_names`.)
* If the sysadmin is more eager to save space, they can browse the tables (`SHOW TABLES LIKE 'snap_civicrm_%'`) and drop any they don't care to keep.

Comments
----------------------------------------

* _MySQL schema_: You don't have to define all the fiddly schema details for each snapshot table - because it builds on `CREATE TABLE ... SELECT ...`, and MySQL is fairly clever about inferring schema for this automatically.
* _Not all deployments_: The PR does not enable snapshots for all deployments.
    * It skips multlingual deployments. I'm sure it's possible -- it's patch/project welcome. I simply haven't looked at it seriously.
    * On very large deployments, MySQL storage can be an issue, and snapshots may aggravate storage challenges. The current PR sets an arbitrary threshold (*200k contacts or 200k contributions or 200k activities...*); if a site exceeds the threshold, then it will NOT create snapshots. It shows a warning instead.
        * I don't think we should expect to reach 100% of sites. I think it's an improvement if we can get snapshots on 80%+ of sites. And the really large systems are likely to have extra constraints that we won't anticipate.
        * I'm not hardset on this particular rule. Maybe some other heuristic or policy-mechanism is better? Is that something we can tune over time?
* _Not a replacement for backups_: One should still take backups for any number of reasons. 
* _Improvements over 5.47_: I think 5.47 is a robust example scenario, but I think this improves on some of the techniques we used/discussed there:
    * _Compared to synthesizing older database dumps from `*.mysql` files_:
        * With automatic `snap_civicrm_*` tables, it's easier to write portable cleanup scripts.
    * _Compared to `civicrm_event.*_bak`  backup columns_:
        * It's easier to define snapshots for additional tables+columns.
        * The phase-out/purging is more clearly defined.
        * At runtime, the snapshot data is "cold". The size/performance/limits of the regular table are demonstrably unaffected.

totten committed:
CRM/Upgrade - Define "Snapshot" helper (with unit-tests)

totten committed:
CRM/Upgrade - Add basic support for snapshots to the upgrade framework

totten committed:
CRM/Upgrade - Enable snapshots for 5.50.alpha1

Note: This is being committed during 5.50.beta1.  If anyone previously
executed the upgrade for the RC/beta/alpha, then may not get snapshots.  But
those upgrades are already done; and they're not guaranteed anyway; and the
snapshots aren't mandatory anyway.

totten committed:
(REF) SnapshotTest - Split big test in two. Cleanup comments.

demeritcowboy commented:
Would these tables be stored forever with trigger-based logging turned on? I can see an argument both ways to keep them or not. Since there's a cleanup task intended to delete them like tmp tables it suggests excluding by default? But I don't have a strong opinion on it.

totten committed:
CRM/Upgrade - Store snapshots with compression. More comments.

totten committed:
CRM/Upgrade - Snapshots should have name `civicrm_snap_{OWNER}_{VERSION}_{NAME}`

totten commented:
>  (@artfulrobot) How about `civicrm_snap_{$COMPONENT}_{$VERSION}_{$NAME}` where `$COMPONENT` would be core or the short name of an extension?

Good point. I was concerned that MySQL had limits on the length of a table-name, but... this can be addressed by inspecting plausible examples. Note the limit is 64 characters.

```
          Tens      Twenties  Thirties  Forties   Fifties   Sixties
123456789 123456789 123456789 123456789 123456789 123456789 1234
-------------------------------------------------------------------
civicrm_snap_v5_61_foobar_whizbang
civicrm_snap_core_v5_61_foobar_whizbang
civicrm_snap_sequentialcreditnotes_v5_61_foobar_whizbang
civicrm_snap_oauth_client_v5_61_foobar_whizbang
```

Even with longish names like `sequentialcreditnotes` and `foobar_whizbang`, there's still ~8 bytes to spare. So it's not a huge amount of spare room... but I think it's ample for most developers to work with.

> I wonder whether this could be easily extended for extensions.

Going back to the general question, I think it's quite possible, although the pre-existing DX is different (`CRM_Upgrade_Incremental_Base` vs `CRM_*_Upgrader`), and each would merit a slightly different sugar-coating (`$this->addSnapshotTask()` vs `???`). IMHO a separate PR would be appropriate for working through the extension POV.

But I think your comment is spot-on in highlighting the naming convention - that part would be annoying to change later. So I'll push up some revisions for `Snapshot` and `SnapshotTest` to show they work with extension-style naming/versioning.

> Perhaps the hard-coded limit of 200k could be made more palletable by at least a constant we can set in civicrm.settings.php? 200k events is really quite a lot, but 200k activities on a long-lived reasonable sized install is not a lot at all.

I'm fine with a constant for now.

There's some tension around configuration-management (eg constant vs setting; on/off-boolean vs threshold-int; one option vs multiple options; config file vs GUI). TBH, I kinda think we should get some basic functionality out there (with some real-world `r-run`) before we make a strong commitment on the config-options.

Perhaps we do a constant with some kind of "experimental" flag on it (ie "this hidden option could be renamed/moved in a future version").

> (@demeritcowboy) Would these tables be stored forever with trigger-based logging turned on? I can see an argument both ways to keep them or not. Since there's a cleanup task intended to delete them like tmp tables it suggests excluding by default? But I don't have a strong opinion on it.

Yeah, I think the trigger-based logging system should totally ignore these tables. I guess the near-term task here is to either `r-run`/patch/confirm that the tables are ignored on logging systems. (Or, if there's some obstacle, just disable snapshots on logged deployments - and revisit in another update.)

totten committed:
CRM/Upgrade - Snapshots should have name `snap_{OWNER}_{VERSION}_{NAME}`

1. Prevents the trigger-based logging system from hitting the snapshots.
2. Makes the de-facto table names shorter (less likely to hit 64 char limit)
3. In practice, it should be easier to skim
4. Use the same pseudo-extension name that we use elsewhere (`civicrm`)

totten commented:
I did some `r-run` with the upgrade-snapshot while also using fine-grained log-tables (`log_civicrm_*`) and confirmed that it created extra tables:

```
mysql> show tables like '%snap%';
+--------------------------------------+
| Tables_in_dmastercivi_pa1rx (%snap%) |
+--------------------------------------+
| civicrm_snap_core_v5_50_fields       |
| civicrm_snap_core_v5_50_mappings     |
| log_civicrm_snap_core_v5_50_fields   |
| log_civicrm_snap_core_v5_50_mappings |
+--------------------------------------+
```

@artfulrobot @demeritcowboy Changing the naming convention again (`snap_{OWNER}_{VERSION}_{NAME}`) seems to be a fairly simple mitigation all around. Example output:

```
mysql> show tables like '%snap%';
+--------------------------------------+
| Tables_in_dmastercivi_pa1rx (%snap%) |
+--------------------------------------+
| snap_civicrm_v5_50_fields            |
| snap_civicrm_v5_50_mappings          |
+--------------------------------------+
```

The convention is still quite distinctive (hard to imagine a conflicted table name), but it has a few benefits:

1. Prevents unintended interaction between the trigger-based logging system and the upgrade-snapshots.
2. Makes the de-facto table names shorter (less likely to hit 64 char limit)
3. In practice, it should be easier to skim (in phpMyAdmin etal)
4. The fake/constant name used for CiviCRM core now matches the fake/constant name used in extensions+resources (ie `civicrm` rather than `core`).

totten committed:
CRM/Upgrade - Add notice about snap_* and MySQL configuration

totten committed:
Define experimental option CIVICRM_UPGRADE_SNAPSHOT

totten commented:
PR updated to address all feedback from @artfulrobot and @demeritcowboy. Rebased.

eileenmcnaughton commented:
I've given this merge ready as feedback appears to be addressed & tests are good

demeritcowboy commented:
On the logging, there's already a spot where some tables are excluded so starting the tablename with `civicrm_` to keep the civi namespacing would still work: https://github.com/civicrm/civicrm-core/blob/51ae64d92c18291070b32d7d6e2a02d8ba280b0a/CRM/Logging/Schema.php#L144. But since they're temporary then whatever works.

totten commented:
> `civicrm_snap_000001` and have the `000001` be a soft FK to an ID in a `civicrm_upgrade_snapshots`

That's interesting. If the limits were shorter (like 16 bytes or 32 bytes), then I could see it. But 64 bytes is quite serviceable (given the audience/use-cases), and there is a complexity-cost to the extra indirection+metadata.

(Don't want to critical, but just to demonstrate a bit -- The table-names would vary from DB-to-DB and that they'd be harder to read/intuit... and it creates edge-cases where the FK breaks... and the table `civicrm_upgrade_snapshots` would itself require periodic maintenance. That sounds normal, but it's really like  `civicrm_queue_item` -- *an upgrade-critical table*. Any maintenance done on this table needs to be a special task that runs *before the main upgrade*.)

I think it's possible to address all that if there's additional metadata that we really need, but I prefer KISS until there's a stronger need.

> ...searchkit could easily be employed to present a list of left over crud...

I think you can get this from `information_schema.tables` and a regex. Here's an implementation of `UpgradeSnapshot::get`

https://gist.github.com/totten/719030a44053fdaf1f408340ae5390e5
